File: src/jsonrpc/eth/getFilterChanges/eth_getFilterChanges.js
/**
 * @fileoverview eth_getFilterChanges JSON-RPC method
 */

import { createRequest } from "../../types/JsonRpcRequest.js";

/**
 * @typedef {import('../../types/index.js').AddressType} Address
 * @typedef {import('../../types/index.js').Hash} Hash
 * @typedef {import('../../types/index.js').Quantity} Quantity
 * @typedef {import('../../types/index.js').BlockTag} BlockTag
 * @typedef {import('../../types/index.js').BlockSpec} BlockSpec
 * @typedef {import('../../types/JsonRpcRequest.js').JsonRpcRequest} JsonRpcRequest
 */

/**
 * Polling method for the filter with the given ID (created using `eth_newFilter`). Returns an array of logs, block hashes, or transaction hashes since last poll, depending on the installed filter.
 *
 * @example
 * Filter identifier: "0x01"
 * Result: ...
 *
 * Implements the `eth_getFilterChanges` JSON-RPC method.
 */

/** The JSON-RPC method name */
export const method = "eth_getFilterChanges";
/**
 * Result for `eth_getFilterChanges`
 *
 * Filter results
 *
 * @typedef {Quantity} Result
 */

/**
 * Request for `eth_getFilterChanges`
 *
 * @typedef {JsonRpcRequest<'eth_getFilterChanges', [Quantity]>} Request
 */

/**
 * Creates a eth_getFilterChanges JSON-RPC request
 *
 * @param {Quantity} address
 * @param {number | string | null} [id] - Optional request ID
 * @returns {Request}
 */
export function GetFilterChangesRequest(address, id = null) {
	return /** @type {Request} */ (createRequest(method, [address], id));
}

File: packages/voltaire-zig/src/jsonrpc/eth/getFilterChanges/eth_getFilterChanges.zig
const std = @import("std");
const types = @import("../../types.zig");

/// Polling method for the filter with the given ID (created using `eth_newFilter`). Returns an array of logs, block hashes, or transaction hashes since last poll, depending on the installed filter.
///
/// Example:
/// Filter identifier: "0x01"
/// Result: ...
///
/// Implements the `eth_getFilterChanges` JSON-RPC method.
pub const EthGetFilterChanges = @This();

/// The JSON-RPC method name
pub const method = "eth_getFilterChanges";

/// Parameters for `eth_getFilterChanges`
pub const Params = struct {
    /// hex encoded unsigned integer
    filter_identifier: types.Quantity,

    pub fn jsonStringify(self: Params, jws: *std.json.Stringify) !void {
        try jws.beginArray();
        try jws.write(self.filter_identifier);
        try jws.endArray();
    }

    pub fn jsonParseFromValue(allocator: std.mem.Allocator, source: std.json.Value, options: std.json.ParseOptions) !Params {
        if (source != .array) return error.UnexpectedToken;
        if (source.array.items.len != 1) return error.InvalidParamCount;

        return Params{
            .filter_identifier = try std.json.innerParseFromValue(types.Quantity, allocator, source.array.items[0], options),
        };
    }
};

/// Result for `eth_getFilterChanges`
pub const Result = struct {
    /// Filter results
    value: types.Quantity,

    pub fn jsonStringify(self: Result, jws: *std.json.Stringify) !void {
        try jws.write(self.value);
    }

    pub fn jsonParseFromValue(allocator: std.mem.Allocator, source: std.json.Value, options: std.json.ParseOptions) !Result {
        return Result{
            .value = try std.json.innerParseFromValue(types.Quantity, allocator, source, options),
        };
    }
};
